File: rust/project/reactions.rb
module RustProjection
  module Projector
    module_function

    # `Policy#event_qualifier`/`#event_name` (`Naming.qualifier`/
    # `Naming.unqualified`, read directly) — NOT separate wire fields
    # (`Policy#to_h` only carries `on_event` whole); the exported
    # canonical IR gives this generator the same raw `on_event` string
    # Ruby's own runtime re-derives these from, so it re-derives them the
    # identical way rather than exporting a fourth field the wire contract
    # doesn't have.
    def policy_event_qualifier(on_event)
      text = on_event.to_s
      text.include?(".") ? text.split(".", 2).first : nil
    end

    def policy_event_name(on_event)
      text = on_event.to_s
      text.include?(".") ? text.split(".", 2).last : text
    end

    # ── THE POLICY TABLE — `kernel::orchestrate`'s own static data,
    # `Runtime::PolicyInterpreter#policies_for`/`#deliver` ported to
    # generated `PolicyRule` rows a hand-written, generic function walks
    # (kernel/orchestrate.rs), the SAME "compile shapes, interpret
    # behavior" split every other generated/kernel pair in this project
    # already holds to.
    #
    # CROSS-DOMAIN POLICIES ARE SPLIT OUT, not skipped: `bin/project_rust
    # <domain>` still compiles exactly ONE target domain (plus the
    # self-hosted meta-language) into one `Store` — a policy whose
    # `across` names a domain THIS build never generated still has no
    # `dispatch_by_name`/`Store` to route to LOCALLY. What changed is
    # that local dispatch is no longer the only way this project delivers
    # a command: `rust/host` (the unsandboxed host layer, real AWS SDK
    # access this WASM-compiled kernel structurally lacks — see
    # orchestrate.rs's own header) can invoke that OTHER domain's own
    # deployed Lambda directly, a straight port of Ruby's own
    # `Adapters::Lambda::Client`. `emit_cross_domain_policy_table`, below,
    # emits these rows into their OWN table (`CrossDomainPolicyRule`, not
    # `PolicyRule` — a different shape, since there is no local
    # `target_verb` to dispatch, only a domain+verb pair for rust/host's
    # `lambda_client.rs` to invoke remotely) instead of dropping them.
    def emit_policy_table(domain_name, policies, aggregates = [])
      rows = local_policy_rows(domain_name, policies, aggregates)

      Exemplar.render(
        "policy_table",
        'crate::kernel::PolicyRule { policy_name: "tmpl_policy_name", event_name: "tmpl_event_name", event_qualifier: None, target_verb: "tmpl_target_verb", for_each: None, for_each_key: None, with_spec: &[] },' => rows.join("\n")
      )
    end

    # ── THE MERGED POLICY TABLE — bin/project_rust's own `merged.rs`,
    # spanning the target domain AND every framework/vendored chapter it
    # attaches, the SAME union `merged_aggregates`/`merged_queries` there
    # already are. RECOVERS A DOCUMENTED, DELIBERATE GAP: bin/project_rust
    # used to call `emit_policy_table` with ONLY the target domain's own
    # policies — "Policies/process managers do NOT merge across chapters
    # here — Governance/Identity declare none today... not attempted by
    # this pass" (that comment's own words). Governance/Identity really
    # don't declare any, so the gap was invisible until a domain vendored
    # a chapter that DOES — found live, generating lifeadelics' vendored
    # embryonaut_bluebooks/payments: `Payments::Payment.PaymentGateway
    # .Succeeded` (a real, generated, correctly-dispatchable port
    # operation — dispatch_by_name already routes it fine) emitted
    # `PaymentConfirmedByProcessor` exactly as declared, but the MERGED
    # Store's own POLICIES table was `&[]` — Payments' 8 policies existed
    # in payments/registry.rs's own STANDALONE table (each chapter still
    # gets its own, unchanged, via emit_policy_table above), just never
    # folded into the ONE table kernel::orchestrate actually reads at
    # runtime. `OnPaymentConfirmedByProcessor`'s own trigger never fired;
    # a real webhook would have left the payment stuck "pending" forever.
    #
    # `sources` — one `{domain_name:, policies:, aggregates:}` per
    # chapter (target first, then each attached one, matching
    # `chapter_mod_names`' own insertion order) — each entry runs through
    # the IDENTICAL `local_policy_rows` a standalone chapter's own table
    # already uses, so a policy's own `target_verb` is qualified against
    # ITS OWN domain_name/aggregates, never the target's.
    def emit_merged_policy_table(sources)
      rows = sources.flat_map { |source| local_policy_rows(source[:domain_name], source[:policies], source[:aggregates]) }

      Exemplar.render(
        "policy_table",
        'crate::kernel::PolicyRule { policy_name: "tmpl_policy_name", event_name: "tmpl_event_name", event_qualifier: None, target_verb: "tmpl_target_verb", for_each: None, for_each_key: None, with_spec: &[] },' => rows.join("\n")
      )
    end

    def local_policy_rows(domain_name, policies, aggregates = [])
      policies.filter_map do |policy|
        target_domain = policy[:target_domain] || domain_name
        next nil unless target_domain == domain_name

        event_name = policy_event_name(policy[:on_event])
        qualifier = policy_event_qualifier(policy[:on_event])
        qualifier_expr = qualifier ? "Some(#{qualifier.inspect})" : "None"
        target_verb = "#{target_domain}::#{policy[:trigger_command]}"

        # `policy_name` — orchestrate.rs's own `reaction_log`/`saga_log`
        # entries need the policy's own declared name (`PolicyInterpreter
        # #deliver`'s `record = { policy: policy.name, ... }`, read
        # directly) — previously absent here on purpose (this struct's own
        # OLD header: "nothing downstream of a same-domain reaction ever
        # needed to name the policy that caused it"), now needed the same
        # way `CrossDomainPolicyRule` already carries it.
        "    crate::kernel::PolicyRule { policy_name: #{policy[:name].to_s.inspect}, event_name: #{event_name.inspect}, " \
          "event_qualifier: #{qualifier_expr}, target_verb: #{target_verb.inspect}, " \
          "for_each: #{fan_out_verb_expr(domain_name, policy)}, " \
          "for_each_key: #{fan_out_key_expr(domain_name, policy, aggregates)}, " \
          "with_spec: #{with_spec_expr(policy)} },"
      end
    end


    # ── `for_each` — THE FAN-OUT'S OWN QUERY, qualified here rather than
    # in the kernel. `Behaviour::Policy#for_each_route` resolves the bare
    # "Aggregate.query" spelling against the policy's OWN domain, and
    # that resolution is a fact about the source: settling it at codegen
    # keeps the kernel's own lookup a plain table hit, the same shape
    # `cli::run` already answers a top-level ask with.
    def fan_out_verb_expr(domain_name, policy)
      for_each = policy[:for_each].to_s
      return "None" if for_each.empty?

      "Some(#{(for_each.include?("::") ? for_each : "#{domain_name}::#{for_each}").inspect})"
    end

    # ── THE NAME A MATCHED ROW'S ID IS MINTED UNDER, which is the TARGET
    # COMMAND'S question and not the aggregate's: a command declared ON
    # the aggregate it references is addressed by that aggregate's own
    # bare reference key, and one merely HOLDING a reference to it is
    # addressed by the attribute that holds it. `Behaviour::Command
    # #addressing_key_for` is the rule, and this is the same rule read
    # off the exported IR — `spec/codegen_parity_spec.rb` holds this
    # generator byte-identical to `rust/codegen`'s own twin, which is
    # what keeps the two spellings of it from drifting.
    def fan_out_key_expr(domain_name, policy, aggregates)
      for_each = policy[:for_each].to_s
      return "None" if for_each.empty?

      path, = for_each.split(".", 2)
      row_aggregate = path.to_s.include?("::") ? path.split("::", 2).last : path
      command = target_command_for(policy, aggregates)
      key = command && addressing_key_for(command, row_aggregate)
      key ? "Some(#{key.to_s.inspect})" : "None"
    end

    def target_command_for(policy, aggregates)
      aggregate_name, command_name = policy[:trigger_command].to_s.split(".", 2)
      aggregate = Array(aggregates).find { |candidate| candidate[:name].to_s == aggregate_name.to_s }
      Array(aggregate && aggregate[:commands]).find { |candidate| candidate[:name].to_s == command_name.to_s }
    end

    def addressing_key_for(command, aggregate_name)
      return snake_case(aggregate_name) if command[:references].to_s == aggregate_name.to_s

      held = Array(command[:attributes]).find { |attribute| attribute[:type].to_s == "Reference<#{aggregate_name}>" }
      held && held[:name]
    end

    def snake_case(name)
      name.to_s.gsub(/([a-z\d])([A-Z])/, '\\1_\\2').gsub(/([A-Z]+)([A-Z][a-z])/, '\\1_\\2').downcase
    end

    # ── `trigger ..., with:` — WHAT THE TRIGGER IS GIVEN. Each binding
    # rides the wire already rendered (`Literal::render`, so a Symbol
    # keeps its leading colon and stays distinguishable from a literal
    # string of the same spelling); the kernel reads that spelling back.
    def with_spec_expr(policy)
      pairs = Array(policy[:with_spec])
      return "&[]" if pairs.empty?

      "&[#{pairs.map { |key, value| "(#{key.to_s.inspect}, #{value.to_s.inspect})" }.join(', ')}]"
    end

    # ── THE CROSS-DOMAIN POLICY TABLE — every policy `local_policy_rows`
    # above filtered OUT, represented instead of dropped. `target_verb`
    # here is FULLY QUALIFIED (`"#{target_domain}::#{trigger_command}"`),
    # the SAME formula `local_policy_rows` uses, above — NOT the bare
    # `policy[:trigger_command]` this used to emit.
    #
    # FOUND LIVE, deploying a real second domain (Compliance) to actually
    # prove cross-domain delivery for the first time: a bare verb refuses
    # with "unknown command" against ANY compiled target, single-chapter
    # or not — `dispatch_by_name`'s own generated match arms are ALWAYS
    # "Domain::Aggregate.Command", the same convention every other verb
    # in this whole system already uses (`kernel::cli::run`'s own
    # top-level step, `RemoteDispatcher#dispatch`'s real cross-Lambda
    # calls, `Adapters::Lambda::Client#dispatch`). The OLD reasoning here
    # ("the Lambda invoked IS the domain, so the verb it receives is
    # already implicitly scoped to it") assumed a target Lambda's own
    # dispatch table could somehow be UNqualified for a single-chapter
    # compile — never actually true; nothing in `rust/project/registry.rb`
    # ever emits an unqualified match arm. Confirmed both ways: compiled
    # `examples/compliance` refuses a bare "AccountFreezeReview.Open" as
    # "unknown command" and accepts "Compliance::AccountFreezeReview.Open"
    # cleanly. This was never caught before because no domain had ever
    # deployed a real second Lambda to receive one of these calls.
    def cross_domain_policy_rows(domain_name, policies)
      policies.filter_map do |policy|
        target_domain = policy[:target_domain] || domain_name
        next nil if target_domain == domain_name

        event_name = policy_event_name(policy[:on_event])
        qualifier = policy_event_qualifier(policy[:on_event])
        qualifier_expr = qualifier ? "Some(#{qualifier.inspect})" : "None"
        target_verb = "#{target_domain}::#{policy[:trigger_command]}"

        "    crate::kernel::CrossDomainPolicyRule { policy_name: #{policy[:name].to_s.inspect}, event_name: #{event_name.inspect}, " \
          "event_qualifier: #{qualifier_expr}, target_domain: #{target_domain.inspect}, target_verb: #{target_verb.inspect} },"
      end
    end

    def emit_cross_domain_policy_table(domain_name, policies)
      rows = cross_domain_policy_rows(domain_name, policies)

      puts "cross-domain policy table: #{rows.size} row(s) — delivered by rust/host's lambda_client.rs, not locally dispatched" if rows.any?

      Exemplar.render(
        "cross_domain_policy_table",
        'crate::kernel::CrossDomainPolicyRule { policy_name: "tmpl_policy_name", event_name: "tmpl_event_name", event_qualifier: None, target_domain: "tmpl_target_domain", target_verb: "tmpl_target_verb" },' =>
          rows.join("\n")
      )
    end

    # ── THE MERGED CROSS-DOMAIN POLICY TABLE — same recovery as
    # emit_merged_policy_table above, same documented gap, same reason
    # (Governance/Identity declare no cross-domain policies either, so
    # this half was equally invisible until a vendored chapter needed
    # it). `sources` — the identical `{domain_name:, policies:,
    # aggregates:}` array `emit_merged_policy_table` takes (aggregates
    # unused here, cross-domain rows need no fan-out addressing key —
    # see cross_domain_policy_rows' own shape).
    def emit_merged_cross_domain_policy_table(sources)
      rows = sources.flat_map { |source| cross_domain_policy_rows(source[:domain_name], source[:policies]) }

      puts "cross-domain policy table: #{rows.size} row(s) — delivered by rust/host's lambda_client.rs, not locally dispatched" if rows.any?

      Exemplar.render(
        "cross_domain_policy_table",
        'crate::kernel::CrossDomainPolicyRule { policy_name: "tmpl_policy_name", event_name: "tmpl_event_name", event_qualifier: None, target_domain: "tmpl_target_domain", target_verb: "tmpl_target_verb" },' =>
          rows.join("\n")
      )
    end

    # A `with:` binding's raw wire spelling — Hecksagain::Literal's, the
    # same one every other to_h-bound literal field rides. `Marks.read` is
    # the exact, already-proven inverse (mutations.rb's
    # `append_field_source` is the identical round trip on a `sets
    # append:` field) — reused rather than re-derived.
    def with_value_parsed(raw)
      Hecksagain::Bluebook::Assembly::Marks.read(raw)
    end

    # A parsed `with:` literal (`Marks.read`'s own output shape for
    # anything that isn't a Symbol — a Hash/String/Integer/Float/Boolean/
    # nil) to a Rust expression BUILDING the equivalent `Json` value —
    # never a `const` literal (`Json` holds `Vec`/`String`, not
    # `const`-constructible in stable Rust), which is why every literal
    # gets its own one-off generated function (`emit_with_value`, below).
    def json_literal_expr(value)
      case value
      when Hash
        fields = value.map { |k, v| "(#{k.to_s.inspect}.to_string(), #{json_literal_expr(v)})" }
        "crate::kernel::Json::Object(vec![#{fields.join(', ')}])"
      when String
        "crate::kernel::Json::Str(#{value.inspect}.to_string())"
      when Integer
        "crate::kernel::Json::int(#{value})"
      when Float
        "crate::kernel::Json::Num(#{value}f64)"
      when true, false
        "crate::kernel::Json::Bool(#{value})"
      when nil
        "crate::kernel::Json::Null"
      else
        raise "unsupported with: literal #{value.inspect} — json_literal_expr doesn't cover this shape"
      end
    end

    # One `with:` binding, resolved to a `WithValue` — a bare Symbol is a
    # RUNTIME reference (`WithValue::Ref`, resolved by
    # `kernel::orchestrate`'s `resolve_with`); anything else is a LITERAL,
    # emitted as its own tiny `fn() -> Json` (`literal_fns` collects these
    # so the caller can splice them in above the table that references
    # them by name).
    def emit_with_value(raw, literal_fns)
      parsed = with_value_parsed(raw)
      return "crate::kernel::WithValue::Ref(#{parsed.to_s.inspect})" if parsed.is_a?(Symbol)

      fn_name = "pm_literal_#{literal_fns.size}"
      literal_fns << Exemplar.render(
        "with_value_literal_fn",
        "tmpl_literal_fn" => fn_name,
        "tmpl_body_placeholder()" => json_literal_expr(parsed)
      )
      "crate::kernel::WithValue::Literal(#{fn_name})"
    end

    def emit_dispatch_spec(spec, literal_fns)
      with_pairs = spec[:with_spec].map { |key, raw| "(#{key.to_s.inspect}, #{emit_with_value(raw, literal_fns)})" }
      "crate::kernel::DispatchSpec { command_name: #{spec[:command_name].inspect}, with: &[#{with_pairs.join(', ')}] }"
    end

    def emit_handler(handler, literal_fns)
      dispatches = handler[:dispatches].map { |d| emit_dispatch_spec(d, literal_fns) }
      "crate::kernel::Handler { event_type: #{handler[:event_type].inspect}, from_state: #{handler[:from_state].inspect}, to_state: #{handler[:to_state].inspect}, dispatches: &[#{dispatches.join(', ')}] }"
    end

    # ── THE PROCESS MANAGER TABLE — `kernel::orchestrate`'s own static
    # data for `SagaInterpreter#advance`/`#unwind`, the same "compile
    # shapes, interpret behavior" split the policy table above holds to.
    # No cross-domain narrowing needed here the way `emit_policy_table`
    # needs one: every `dispatch` spec's `command_name` is ALREADY fully
    # domain-qualified on the wire (`Banking::Account.Debit`, not
    # `Account.Debit`) — a process manager cannot name a target this
    # single-domain `Store` doesn't hold without that verb simply failing
    # to route at `dispatch_by_name`'s own `unknown command` branch, the
    # same swallowed-refusal path any other unreachable target takes.
    def emit_process_manager_table(process_managers)
      literal_fns = []
      pm_exprs = process_managers.map do |pm|
        handlers = pm[:handlers].map { |h| emit_handler(h, literal_fns) }
        "crate::kernel::ProcessManagerDef { name: #{pm[:name].inspect}, correlates_by: #{pm[:correlates_by].inspect}, " \
          "starts_on: #{pm[:starts_on].inspect}, ends_on: #{pm[:ends_on].inspect}, initial_state: #{pm[:states].first.inspect}, " \
          "handlers: &[#{handlers.join(', ')}] }"
      end

      Exemplar.render(
        "process_manager_table",
        "fn tmpl_literal_fns_placeholder() {}" => literal_fns.join("\n"),
        '    crate::kernel::ProcessManagerDef { name: "tmpl_pm_name", correlates_by: "tmpl_correlates_by", starts_on: "tmpl_starts_on", ends_on: "tmpl_ends_on", initial_state: "tmpl_initial_state", handlers: &[] },' =>
          pm_exprs.map { |e| "    #{e}," }.join("\n")
      )
    end

    # `Naming.reference_key(event.aggregate)`, precomputed per aggregate
    # rather than re-derived at runtime — `Correlation#saga_correlation`'s
    # own THIRD tier (this file's own header on `orchestrate.rs`), needed
    # for real: a leg dispatched WITH its own `reference_to` argument
    # (`Transfer.Credited`'s own `transfer:`, `json_codec.rb`'s
    # `emit_extract_id` header) announces an event whose payload carries
    # THAT key, not `correlates_by`'s own dotted field — `TransferCredited`
    # only ever has `{"transfer": "xfer-1"}`, never `{"reference": {...}}`.
    # `kernel::orchestrate`'s `correlation_of` falls back to this table,
    # keyed by the emitting event's OWN qualified aggregate name, exactly
    # the same per-event (not per-process-manager) rule Ruby's own
    # `Naming.reference_key(event.aggregate)` applies.
    # `chapters` — `[[domain_name, aggregate_names], ...]`, one pair per
    # chapter (bin/project_rust's own per-domain call for a single-
    # chapter registry.rs passes exactly one pair; the merged registry
    # spanning every chapter a domain attaches passes one pair per
    # chapter) — this table is domain-qualified-name -> key regardless
    # of how many chapters fed it, so merging is just "more pairs in
    # the same flat list," no other change needed.
    def emit_reference_key_table(chapters)
      arms = chapters.flat_map do |domain_name, aggregate_names|
        aggregate_names.map do |name|
          qualified = "#{domain_name}::#{name}"
          key = Hecksagain::Naming.snake(name)
          "        #{qualified.inspect} => Some(#{key.inspect}),"
        end
      end

      Exemplar.render("reference_key_table", '"tmpl_qualified" => Some("tmpl_key"),' => arms.join("\n"))
    end
  end
end

File: rust/codegen/src/reactions.rs
//! Port of `rust/project/reactions.rb` — read that file's own header
//! comments in full; this mirrors its algorithm directly, function for
//! function.

use crate::exemplar::Exemplar;
use crate::json::Json;
use crate::literal::{self, Literal};
use crate::naming;

pub fn policy_event_qualifier(on_event: &str) -> Option<String> {
    if on_event.contains('.') {
        Some(on_event.splitn(2, '.').next().unwrap_or("").to_string())
    } else {
        None
    }
}

pub fn policy_event_name(on_event: &str) -> String {
    if on_event.contains('.') {
        on_event.splitn(2, '.').nth(1).unwrap_or("").to_string()
    } else {
        on_event.to_string()
    }
}

/// ── THE POLICY TABLE.
pub fn emit_policy_table(exemplar: &Exemplar, domain_name: &str, policies: &[Json], aggregates: &[Json]) -> String {
    let rows = local_policy_rows(domain_name, policies, aggregates);
    exemplar.render(
        "policy_table",
        &[("crate::kernel::PolicyRule { policy_name: \"tmpl_policy_name\", event_name: \"tmpl_event_name\", event_qualifier: None, target_verb: \"tmpl_target_verb\", for_each: None, for_each_key: None, with_spec: &[] },", rows.join("\n"))],
    )
}

fn local_policy_rows(domain_name: &str, policies: &[Json], aggregates: &[Json]) -> Vec<String> {
    policies
        .iter()
        .filter_map(|policy| {
            let target_domain = policy.get("target_domain").map(Json::to_s).unwrap_or_else(|| domain_name.to_string());
            if target_domain != domain_name {
                return None;
            }

            let on_event = policy.get("on_event").map(Json::to_s).unwrap_or_default();
            let event_name = policy_event_name(&on_event);
            let qualifier = policy_event_qualifier(&on_event);
            let qualifier_expr = match qualifier {
                Some(q) => format!("Some({})", naming::ruby_inspect_string(&q)),
                None => "None".to_string(),
            };
            let target_verb = format!("{target_domain}::{}", policy.get("trigger_command").map(Json::to_s).unwrap_or_default());
            let name = policy.get("name").map(Json::to_s).unwrap_or_default();

            Some(format!(
                "    crate::kernel::PolicyRule {{ policy_name: {}, event_name: {}, event_qualifier: {qualifier_expr}, target_verb: {}, for_each: {}, for_each_key: {}, with_spec: {} }},",
                naming::ruby_inspect_string(&name),
                naming::ruby_inspect_string(&event_name),
                naming::ruby_inspect_string(&target_verb),
                fan_out_verb_expr(domain_name, policy),
                fan_out_key_expr(policy, aggregates),
                with_spec_expr(policy)
            ))
        })
        .collect()
}

/// ── THE CROSS-DOMAIN POLICY TABLE.
pub fn emit_cross_domain_policy_table(exemplar: &Exemplar, domain_name: &str, policies: &[Json]) -> String {
    let rows: Vec<String> = policies
        .iter()
        .filter_map(|policy| {
            let target_domain = policy.get("target_domain").map(Json::to_s).unwrap_or_else(|| domain_name.to_string());
            if target_domain == domain_name {
                return None;
            }

            let on_event = policy.get("on_event").map(Json::to_s).unwrap_or_default();
            let event_name = policy_event_name(&on_event);
            let qualifier = policy_event_qualifier(&on_event);
            let qualifier_expr = match qualifier {
                Some(q) => format!("Some({})", naming::ruby_inspect_string(&q)),
                None => "None".to_string(),
            };
            let target_verb = format!("{target_domain}::{}", policy.get("trigger_command").map(Json::to_s).unwrap_or_default());
            let name = policy.get("name").map(Json::to_s).unwrap_or_default();

            Some(format!(
                "    crate::kernel::CrossDomainPolicyRule {{ policy_name: {}, event_name: {}, event_qualifier: {qualifier_expr}, target_domain: {}, target_verb: {} }},",
                naming::ruby_inspect_string(&name),
                naming::ruby_inspect_string(&event_name),
                naming::ruby_inspect_string(&target_domain),
                naming::ruby_inspect_string(&target_verb)
            ))
        })
        .collect();

    if !rows.is_empty() {
        println!("cross-domain policy table: {} row(s) — delivered by rust/host's lambda_client.rs, not locally dispatched", rows.len());
    }

    exemplar.render(
        "cross_domain_policy_table",
        &[(
            "crate::kernel::CrossDomainPolicyRule { policy_name: \"tmpl_policy_name\", event_name: \"tmpl_event_name\", event_qualifier: None, target_domain: \"tmpl_target_domain\", target_verb: \"tmpl_target_verb\" },",
            rows.join("\n"),
        )],
    )
}

fn with_value_parsed(raw: &str) -> Literal {
    literal::read(raw)
}

/// A parsed `with:` literal to a Rust expression BUILDING the equivalent
/// `Json` value.
fn json_literal_expr(value: &Literal) -> String {
    match value {
        Literal::Hash(pairs) => {
            let fields: Vec<String> = pairs.iter().map(|(k, v)| format!("({}.to_string(), {})", naming::ruby_inspect_string(k), json_literal_expr(v))).collect();
            format!("crate::kernel::Json::Object(vec![{}])", fields.join(", "))
        }
        Literal::Str(s) => format!("crate::kernel::Json::Str({}.to_string())", naming::ruby_inspect_string(s)),
        Literal::Int(n) => format!("crate::kernel::Json::int({n})"),
        Literal::Float(n) => format!("crate::kernel::Json::Num({}f64)", ruby_float_text(*n)),
        Literal::Bool(b) => format!("crate::kernel::Json::Bool({b})"),
        Literal::Nil => "crate::kernel::Json::Null".to_string(),
        other => panic!("unsupported with: literal {other:?} — json_literal_expr doesn't cover this shape"),
    }
}

fn ruby_float_text(n: f64) -> String {
    let text = format!("{n}");
    if text.contains('.') || text.contains('e') || text.contains('E') {
        text
    } else {
        format!("{text}.0")
    }
}

fn emit_dispatch_spec(exemplar: &Exemplar, spec: &Json, literal_fns: &mut Vec<String>) -> String {
    let with_spec = spec.get("with_spec").map(Json::each).unwrap_or(&[]);
    let with_pairs: Vec<String> = with_spec
        .iter()
        .map(|pair| {
            let kv = pair.as_array().unwrap_or(&[]);
            let key = kv.first().map(Json::to_s).unwrap_or_default();
            let raw = kv.get(1).map(Json::to_s).unwrap_or_default();
            format!("({}, {})", naming::ruby_inspect_string(&key), emit_with_value(exemplar, &raw, literal_fns))
        })
        .collect();
    format!("crate::kernel::DispatchSpec {{ command_name: {}, with: &[{}] }}", naming::ruby_inspect_string(&spec.get("command_name").map(Json::to_s).unwrap_or_default()), with_pairs.join(", "))
}

/// One `with:` binding, resolved to a `WithValue` — a bare Symbol is a
/// RUNTIME reference; anything else is a LITERAL, emitted as its own tiny
/// `fn() -> Json` (`literal_fns` collects these).
fn emit_with_value(exemplar: &Exemplar, raw: &str, literal_fns: &mut Vec<String>) -> String {
    let parsed = with_value_parsed(raw);
    if let Literal::Symbol(name) = &parsed {
        return format!("crate::kernel::WithValue::Ref({})", naming::ruby_inspect_string(name));
    }

    let fn_name = format!("pm_literal_{}", literal_fns.len());
    literal_fns.push(exemplar.render("with_value_literal_fn", &[("tmpl_literal_fn", fn_name.clone()), ("tmpl_body_placeholder()", json_literal_expr(&parsed))]));
    format!("crate::kernel::WithValue::Literal({fn_name})")
}

fn emit_handler(exemplar: &Exemplar, handler: &Json, literal_fns: &mut Vec<String>) -> String {
    let dispatches = handler.get("dispatches").map(Json::each).unwrap_or(&[]);
    let dispatch_strs: Vec<String> = dispatches.iter().map(|d| emit_dispatch_spec(exemplar, d, literal_fns)).collect();
    format!(
        "crate::kernel::Handler {{ event_type: {}, from_state: {}, to_state: {}, dispatches: &[{}] }}",
        naming::ruby_inspect_string(&handler.get("event_type").map(Json::to_s).unwrap_or_default()),
        naming::ruby_inspect_string(&handler.get("from_state").map(Json::to_s).unwrap_or_default()),
        naming::ruby_inspect_string(&handler.get("to_state").map(Json::to_s).unwrap_or_default()),
        dispatch_strs.join(", ")
    )
}

/// ── THE PROCESS MANAGER TABLE.
pub fn emit_process_manager_table(exemplar: &Exemplar, process_managers: &[Json]) -> String {
    let mut literal_fns: Vec<String> = Vec::new();
    let pm_exprs: Vec<String> = process_managers
        .iter()
        .map(|pm| {
            let handlers = pm.get("handlers").map(Json::each).unwrap_or(&[]);
            let handler_strs: Vec<String> = handlers.iter().map(|h| emit_handler(exemplar, h, &mut literal_fns)).collect();
            let states = pm.get("states").map(Json::each).unwrap_or(&[]);
            let initial_state = states.first().map(Json::to_s).unwrap_or_default();
            format!(
                "crate::kernel::ProcessManagerDef {{ name: {}, correlates_by: {}, starts_on: {}, ends_on: {}, initial_state: {}, handlers: &[{}] }}",
                naming::ruby_inspect_string(&pm.get("name").map(Json::to_s).unwrap_or_default()),
                naming::ruby_inspect_string(&pm.get("correlates_by").map(Json::to_s).unwrap_or_default()),
                naming::ruby_inspect_string(&pm.get("starts_on").map(Json::to_s).unwrap_or_default()),
                naming::ruby_inspect_string(&pm.get("ends_on").map(Json::to_s).unwrap_or_default()),
                naming::ruby_inspect_string(&initial_state),
                handler_strs.join(", ")
            )
        })
        .collect();

    exemplar.render(
        "process_manager_table",
        &[
            ("fn tmpl_literal_fns_placeholder() {}", literal_fns.join("\n")),
            (
                "    crate::kernel::ProcessManagerDef { name: \"tmpl_pm_name\", correlates_by: \"tmpl_correlates_by\", starts_on: \"tmpl_starts_on\", ends_on: \"tmpl_ends_on\", initial_state: \"tmpl_initial_state\", handlers: &[] },",
                pm_exprs.iter().map(|e| format!("    {e},")).collect::<Vec<_>>().join("\n"),
            ),
        ],
    )
}

/// `Naming.reference_key(event.aggregate)`, precomputed per aggregate.
pub fn emit_reference_key_table(exemplar: &Exemplar, chapters: &[(String, Vec<String>)]) -> String {
    let arms: Vec<String> = chapters
        .iter()
        .flat_map(|(domain_name, aggregate_names)| {
            aggregate_names.iter().map(move |name| {
                let qualified = format!("{domain_name}::{name}");
                let key = crate::hecksagain_naming::snake(name);
                format!("        {} => Some({}),", naming::ruby_inspect_string(&qualified), naming::ruby_inspect_string(&key))
            })
        })
        .collect();

    exemplar.render("reference_key_table", &[("\"tmpl_qualified\" => Some(\"tmpl_key\"),", arms.join("\n"))])
}

/// `for_each` — THE FAN-OUT'S OWN QUERY, qualified here rather than in
/// the kernel: `Behaviour::Policy#for_each_route` resolves the bare
/// "Aggregate.query" spelling against the policy's own domain, and that
/// resolution is a fact about the source.
fn fan_out_verb_expr(domain_name: &str, policy: &Json) -> String {
    let for_each = policy.get("for_each").map(Json::to_s).unwrap_or_default();
    if for_each.is_empty() {
        return "None".to_string();
    }
    let qualified = if for_each.contains("::") { for_each } else { format!("{domain_name}::{for_each}") };
    format!("Some({})", naming::ruby_inspect_string(&qualified))
}

/// THE NAME A MATCHED ROW'S ID IS MINTED UNDER — the TARGET COMMAND'S
/// question, not the aggregate's. `Behaviour::Command#addressing_key_for`
/// is the rule; this is that rule read off the exported IR, and
/// `spec/codegen_parity_spec.rb` holds it byte-identical to
/// `rust/project/reactions.rb`'s own spelling of the same thing.
fn fan_out_key_expr(policy: &Json, aggregates: &[Json]) -> String {
    let for_each = policy.get("for_each").map(Json::to_s).unwrap_or_default();
    if for_each.is_empty() {
        return "None".to_string();
    }
    let path = for_each.split('.').next().unwrap_or_default().to_string();
    let row_aggregate = path.rsplit("::").next().unwrap_or(&path).to_string();

    let trigger = policy.get("trigger_command").map(Json::to_s).unwrap_or_default();
    let mut parts = trigger.splitn(2, '.');
    let target_aggregate = parts.next().unwrap_or_default().to_string();
    let target_command = parts.next().unwrap_or_default().to_string();

    let key = aggregates
        .iter()
        .find(|a| a.get("name").map(Json::to_s).unwrap_or_default() == target_aggregate)
        .and_then(|a| a.get("commands"))
        .and_then(|commands| match commands {
            Json::Array(items) => items
                .iter()
                .find(|c| c.get("name").map(Json::to_s).unwrap_or_default() == target_command)
                .cloned(),
            _ => None,
        })
        .and_then(|command| addressing_key_for(&command, &row_aggregate));

    match key {
        Some(name) => format!("Some({})", naming::ruby_inspect_string(&name)),
        None => "None".to_string(),
    }
}

fn addressing_key_for(command: &Json, aggregate_name: &str) -> Option<String> {
    if command.get("references").map(Json::to_s).unwrap_or_default() == aggregate_name {
        return Some(crate::hecksagain_naming::snake(aggregate_name));
    }

    let wanted = format!("Reference<{aggregate_name}>");
    match command.get("attributes") {
        Some(Json::Array(items)) => items
            .iter()
            .find(|a| a.get("type").map(Json::to_s).unwrap_or_default() == wanted)
            .and_then(|a| a.get("name").map(Json::to_s)),
        _ => None,
    }
}

/// `trigger ..., with:` — each binding already rendered on the wire
/// (`Literal::render`, so a Symbol keeps its leading colon).
fn with_spec_expr(policy: &Json) -> String {
    let pairs = match policy.get("with_spec") {
        Some(Json::Array(items)) => items.clone(),
        _ => Vec::new(),
    };
    if pairs.is_empty() {
        return "&[]".to_string();
    }
    let rendered: Vec<String> = pairs
        .iter()
        .filter_map(|pair| match pair {
            Json::Array(kv) if kv.len() == 2 => Some(format!(
                "({}, {})",
                naming::ruby_inspect_string(&kv[0].to_s()),
                naming::ruby_inspect_string(&kv[1].to_s())
            )),
            _ => None,
        })
        .collect();
    format!("&[{}]", rendered.join(", "))
}
